### PR TITLE
Add locking to generation of master and media playlists

### PR DESCRIPTION
**Pull request by darkdarkdragon**
Add locks around generation of playlists

### CODE REVIEW

File: writer.go [hunk 1]
```diff
@@ -65,18 +69,23 @@ func (p *MasterPlaylist) Append(uri string, chunklist *MediaPlaylist, params Var
...
 	}
-	p.buf.Reset()
+	p.dirty = true
 }
 
 func (p *MasterPlaylist) ResetCache() {
-	p.buf.Reset()
+	p.lock.Lock()
+	p.dirty = true
+	p.lock.Unlock()
```
Comment (mk-livepeer) [inline writer.go]: For code consistency, please `p.lock.Lock()` followed by `defer p.lock.Unlock()`

File: writer.go [hunk 2]
```diff
@@ -242,6 +251,7 @@ func (p *MasterPlaylist) Version() uint8 {
 // automatically by other Set methods.
 func (p *MasterPlaylist) SetVersion(ver uint8) {
 	p.ver = ver
+	p.dirty = true
```
Comment (mk-livepeer) [inline writer.go:254]: Does this need to be protected by the lock ?

File: writer.go [hunk 3]
```diff
@@ -257,6 +267,7 @@ func (p *MasterPlaylist) String() string {
 func NewMediaPlaylist(winsize uint, capacity uint) (*MediaPlaylist, error) {
 	p := new(MediaPlaylist)
 	p.ver = minver
+	p.dirty = true
```
Comment (mk-livepeer) [inline writer.go:270]: Does this need to be protected by the lock ?

File: writer.go [hunk 4]
```diff
@@ -278,6 +289,13 @@ func (p *MediaPlaylist) last() uint {
 // Remove current segment from the head of chunk slice form a media playlist. Useful for sliding playlists.
 // This operation does reset playlist cache.
 func (p *MediaPlaylist) Remove() (err error) {
+	p.lock.Lock()
+	err = p.remove()
+	p.lock.Unlock()
```
Comment (mk-livepeer) [inline writer.go]: For code consistency, please p.lock.Lock() followed by defer p.lock.Unlock()

File: writer.go [hunk 5]
```diff
@@ -286,24 +304,36 @@ func (p *MediaPlaylist) Remove() (err error) {
 	if p.Live {
 		p.SeqNo++
 	}
-	p.buf.Reset()
+	p.dirty = true
```
Comment (mk-livepeer) [inline writer.go:307]: Does this need to be protected by the lock ?

File: writer.go [hunk 6]
```diff
@@ -313,7 +343,7 @@ func (p *MediaPlaylist) AppendSegment(seg *MediaSegment) error {
 	if p.TargetDuration < seg.Duration {
 		p.TargetDuration = math.Ceil(seg.Duration)
 	}
-	p.buf.Reset()
+	p.dirty = true
```
Comment (mk-livepeer) [inline writer.go:346]: Does this need to be protected by the lock ?

File: writer.go [hunk 7]
```diff
@@ -401,24 +431,31 @@ func (p *MediaPlaylist) InsertSegment(seqNo uint64, seg *MediaSegment) error {
...
-	p.Append(uri, duration, title)
+	p.append(uri, duration, title)
 }
 
 // Reset playlist cache. Next called Encode() will regenerate playlist from the chunk slice.
 func (p *MediaPlaylist) ResetCache() {
-	p.buf.Reset()
+	p.lock.Lock()
+	p.dirty = true
+	p.lock.Unlock()
```
Comment (mk-livepeer) [inline writer.go]: For code consistency, please p.lock.Lock() followed by defer p.lock.Unlock()

File: structure.go [hunk 8]
```diff
@@ -123,6 +123,7 @@ type MediaPlaylist struct {
 	Map            *Map // EXT-X-MAP is optional tag specifies how to obtain the Media Initialization Section (default map for the playlist)
 	WV             *WV  // Widevine related tags outside of M3U8 specs
 	lock           *sync.Mutex
+	dirty          bool
```
Comment (mk-livepeer) [inline structure.go]: please add a comment indicating what elements this lock is protecting

File: structure.go [hunk 9]
```diff
@@ -146,6 +147,8 @@ type MasterPlaylist struct {
 	CypherVersion string // non-standard tag for Widevine (see also WV struct)
 	buf           bytes.Buffer
 	ver           uint8
+	lock          *sync.Mutex
+	dirty         bool
```
Comment (mk-livepeer) [inline structure.go]: please add a comment indicating what elements this lock is protecting

**Review by mk-livepeer — COMMENTED** [review]
Mostly questions as to the coverage of the lock and indication of what elements it is protecting.   I am requesting for consistent usage of the lock / defer unlock, mainly for the sake of future development.

It's best to stick with one pattern (reciprocal balancing of lock with deferred unlock) unless there is a specific reason otherwise.   I generally prefer using a go routine to range over a channel of closures to handle critical sections like this, as opposed to mutex primitives, but there is nothing wrong with this approach.  In fact, in this case, the use of mutex primitives seems to fit nicely.

File: writer.go (hunk 6 above)
Comment (darkdarkdragon) [reply writer.go:346]: No, it's locked in the calling function.

File: writer.go (hunk 5 above)
Comment (darkdarkdragon) [reply writer.go:307]: No, it is locked in the calling function

File: writer.go (hunk 3 above)
Comment (darkdarkdragon) [reply writer.go:270]: No, it is constructor.

File: writer.go (hunk 2 above)
Comment (darkdarkdragon) [reply writer.go:254]: No, nothing can broke here if left unprotected.

File: writer.go (hunk 1 above)
Comment (darkdarkdragon) [reply writer.go]: fixed

File: writer.go (hunk 4 above)
Comment (darkdarkdragon) [reply writer.go]: fixed

File: writer.go (hunk 7 above)
Comment (darkdarkdragon) [reply writer.go]: fixed

File: structure.go (hunk 8 above)
Comment (darkdarkdragon) [reply structure.go]: fixed

File: structure.go (hunk 9 above)
Comment (darkdarkdragon) [reply structure.go]: fixed

**Comment by darkdarkdragon** [comment]
@mklmnop fixed/commented all the things, you can look again

**Review by mk-livepeer — APPROVED** [review]
approved 👍

**Comment by ericxtang** [comment]
Thanks @mk-livepeer

File: writer.go (hunk 3 above)
Comment (ericxtang) [reply writer.go:270]: Maybe I was a little too trigger-happy when I merged the code.  Is this comment resolved?  @mk-livepeer I'm not sure if I understand the comment here - there isn't a lock statement in this block of code.  Can you explain more?

Comment (mk-livepeer) [reply writer.go:270]: @ericxtang As I am still new to the codebase, I raised this comment as a question.  "Does this need to be protected by the lock?"  ...  Ivan explained that this particular function is a constructor -- this function does not need to be protected by the lock - all is well.

**Pull request merged**